### PR TITLE
[SPARK-24234][SS] Support multiple row writers in continuous processing shuffle reader.

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

https://docs.google.com/document/d/1IL4kJoKrZWeyIhklKUJqsW-yEN7V7aL05MmM65AYOfE/edit#heading=h.8t3ci57f7uii

Support multiple different row writers in continuous processing shuffle reader.

Note that having multiple read-side buffers ended up being the natural way to do this. Otherwise it's hard to express the constraint of sending an epoch marker only when all writers have sent one.

## How was this patch tested?

new unit tests

### CODE REVIEW

**Comment by jose-torres** [comment]
@HeartSaVioR @arunmahadevan @xuanyuanking @tdas

**Comment by jose-torres** [comment]
In the above run Zookeeper couldn't get set up for some reason. It's surely unrelated, we aren't touching any currently executed codepath here.

**Comment by jose-torres** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala [hunk 1]
```diff
@@ -115,17 +124,36 @@ class ContinuousShuffleReadSuite extends StreamTest {
     assert(secondEpoch.toSeq.map(_.getInt(0)) == Seq(222, 333))
   }
 
+  test("multiple writers") {
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: nit: move this below with all the other multi writer tests.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala [hunk 2]
```diff
@@ -181,4 +211,80 @@ class ContinuousShuffleReadSuite extends StreamTest {
...
+    // After checking the right rows, block until we get an epoch marker indicating there's no next.
+    // (Also fail the assertion if for some reason we get a row.)
+    val readEpochMarkerThread = new Thread {
+      override def run(): Unit = {
+        assert(!epoch.hasNext)
+      }
+    }
+
+    readEpochMarkerThread.start()
+
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: nit: remove empty line.

**Review by tdas — CHANGES_REQUESTED** [review]
Mostly looks good. I like the user of ExecutorCompletionService. My only concern is the recursive call structured inside the NextIterator. I think its better to convert to a while loop, for multiple reason as commented below.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala [hunk 3]
```diff
@@ -161,13 +189,15 @@ class ContinuousShuffleReadSuite extends StreamTest {
...
     val rdd = new ContinuousShuffleReadRDD(sparkContext, numPartitions = 1)
+    val epoch = rdd.compute(rdd.partitions(0), ctx)
 
     val readRowThread = new Thread {
       override def run(): Unit = {
-        // set the non-inheritable thread local
-        TaskContext.setTaskContext(ctx)
-        val epoch = rdd.compute(rdd.partitions(0), ctx)
-        epoch.next().getInt(0)
+        try {
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: I realized when writing a test further down that it was cleaner to get the RDD outside the thread, and that not catching the InterruptedException was spamming my console.

[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: okay. that make sense. I was worried it was for some nasty error condition that i missed in the previous PR review.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 4]
```diff
@@ -56,20 +62,46 @@ private[shuffle] class UnsafeRowReceiver(
...
-      override def close(): Unit = {}
+      (0 until numShuffleWriters).foreach(writerId => completion.submit(completionTask(writerId)))
+
+      override def getNext(): UnsafeRow = {
+        completion.take().get() match {
+          case ReceiverRow(writerId, r) =>
+            // Start reading the next element in the queue we just took from.
+            completion.submit(completionTask(writerId))
+            r
+            // TODO use writerId
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: It was for the verification you mentioned below. I left it as a marker for myself and then forgot to go back to it :(

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 5]
```diff
@@ -56,20 +62,46 @@ private[shuffle] class UnsafeRowReceiver(
...
+          case ReceiverEpochMarker(writerId) =>
+            // Don't read any more from this queue. If all the writers have sent epoch markers,
+            // the epoch is over; otherwise we need rows from one of the remaining writers.
+            val writersCompleted = numWriterEpochMarkers.incrementAndGet()
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: A map ended up being a more natural way to express the logic I think.

**Comment by jose-torres** [comment]
addressed comments

**Review by HeartSaVioR — COMMENTED** [review]
LGTM, with minor suggestion.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 6]
```diff
@@ -56,20 +69,73 @@ private[shuffle] class UnsafeRowReceiver(
...
-      override def getNext(): UnsafeRow = queue.take() match {
-        case ReceiverRow(r) => r
-        case ReceiverEpochMarker() =>
-          finished = true
-          null
+      // An array of flags for whether each writer ID has gotten an epoch marker.
+      private val writerEpochMarkersReceived =
+        mutable.Map.empty[Int, Boolean].withDefaultValue(false)
+
+      private val executor = Executors.newFixedThreadPool(numShuffleWriters)
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: And I'm also now seeing this approach as alternative to deal with alignment (not buffer rows explicitly but just don't read after epoch comes in). Nice approach.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 7]
```diff
@@ -41,11 +50,15 @@ private[shuffle] case class ReceiverEpochMarker() extends UnsafeRowReceiverMessa
...
 private[shuffle] class UnsafeRowReceiver(
       queueSize: Int,
+      numShuffleWriters: Int,
+      checkpointIntervalMs: Long,
       override val rpcEnv: RpcEnv)
     extends ThreadSafeRpcEndpoint with ContinuousShuffleReader with Logging {
   // Note that this queue will be drained from the main task thread and populated in the RPC
   // response thread.
-  private val queue = new ArrayBlockingQueue[UnsafeRowReceiverMessage](queueSize)
+  private val queues = Array.fill(numShuffleWriters) {
```
Comment (arunmahadevan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Queue per writer makes it easier to handle the alignment. However, if the queues do not get filled evenly, it makes it much more difficult to set a single size across all the queues. In a single queue this would not be much of an issue since the queues are shared by all writers.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 8]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
     new NextIterator[UnsafeRow] {
-      override def getNext(): UnsafeRow = queue.take() match {
-        case ReceiverRow(r) => r
-        case ReceiverEpochMarker() =>
-          finished = true
-          null
+      // An array of flags for whether each writer ID has gotten an epoch marker.
+      private val writerEpochMarkersReceived = Array.fill(numShuffleWriters)(false)
+
+      private val executor = Executors.newFixedThreadPool(numShuffleWriters)
```
Comment (arunmahadevan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Are we creating threads equal to num writers here and do we need it? I assume there would be already the RDD task thread per partition that's invoking `getNext` and cant we return the row from the right queue(s) there ? My concern is that these threads are going to bloat up the total threads to m*n (rdd partitions * shuffle writers) and this may not be very scalable considering the amount of context switch.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala (hunk 7 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Well, is it a problem if we don't set a single size across all the queues?

Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Then it would be configuration nightmare..It should be ok to start with the proposed approach and use same size for all the queues. We can change it later if we see issues as long as the internal details are not exposed to the user.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala (hunk 8 above)
Comment (jose-torres) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: I tried to make that work, but I couldn't figure out how to determine from the task thread which queue is the right one. We can't round-robin, or else we'll fail to make progress if some writers send more data than others.

As mentioned in the doc, this iteration of the shuffle functionality isn't meant to be particularly scalable - we're already incurring a bunch of overhead from using the RPC framework to send data around. The focus for now is correctness and debuggability. Once we have everything in place, we'll swap out everything between ContinuousShuffleReader and ContinuousShuffleWriter with some sort of TCP socket approach, which will probably be able to do away with the N queues entirely by leveraging TCP backpressure.

Comment (arunmahadevan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Ah, yes the n-queues makes it difficult. It would have been straight forward if there was only one queue. I guess here you would have to round-robin and poll the queues with very small timeouts so that you don't block on any queue. If this is a prototype either approach is fine.

**Comment by jose-torres** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala (hunk 7 above)
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: I agree. This level of issues can be dealt with later once we run this and observed the issues for real. Also, it's likely that we will eventually replace a whole lot of this with a non-RPC-endpoint-based transfer mechanism. So lets keep something very simple for now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/ContinuousShuffleReadRDD.scala [hunk 9]
```diff
@@ -42,16 +47,24 @@ case class ContinuousShuffleReadPartition(index: Int, queueSize: Int) extends Pa
...
+ * @param numShuffleWriters the number of continuous shuffle writers feeding into this RDD
+ * @param checkpointIntervalMs the checkpoint interval of the streaming query
  */
 class ContinuousShuffleReadRDD(
     sc: SparkContext,
     numPartitions: Int,
-    queueSize: Int = 1024)
+    queueSize: Int = 1024,
+    numShuffleWriters: Int = 1,
+    checkpointIntervalMs: Long = 1000)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/ContinuousShuffleReadRDD.scala]: Isnt this the same as epochInterval? the term "epoch" is more well known in the code than "checkpoint"

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 10]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
+       */
+      override def getNext(): UnsafeRow = {
+        var nextRow: UnsafeRow = null
+        while (nextRow == null) {
+          nextRow = completion.poll(checkpointIntervalMs, TimeUnit.MILLISECONDS) match {
+            case null =>
+              // Try again if the poll didn't wait long enough to get a real result.
+              // But we should be getting at least an epoch marker every checkpoint interval.
+              logWarning(
+                s"Completion service failed to make progress after $checkpointIntervalMs ms")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Maybe also print which writers (and how many) are we waiting on, so that we can debug.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 11]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
+       * In each call to getNext(), we pull the next row available in the completion queue, and then
+       * submit another task to read the next row from the writer which returned it.
+       *
+       * When a writer sends an epoch marker, we note that it's finished and don't submit another
+       * task for it in this epoch. The iterator is over once all writers have sent an epoch marker.
+       */
+      override def getNext(): UnsafeRow = {
+        var nextRow: UnsafeRow = null
+        while (nextRow == null) {
+          nextRow = completion.poll(checkpointIntervalMs, TimeUnit.MILLISECONDS) match {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: move this inside... its hard to follow the pattern `x = // some large code structured`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 12]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
+                completion.submit(completionTask(writerId))
+                r
+              case ReceiverEpochMarker(writerId) =>
+                // Don't read any more from this queue. If all the writers have sent epoch markers,
+                // the epoch is over; otherwise we need to poll from the remaining writers.
+                writerEpochMarkersReceived(writerId) = true
+                if (writerEpochMarkersReceived.forall(flag => flag)) {
+                  finished = true
+                  // Break out of the while loop and end the iterator.
+                  return null
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: super nit: I personally find these sort of escape hard to read in code. Consider making a separate flag to end the while loop.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 13]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
+            case future => future.get() match {
+              case ReceiverRow(writerId, r) =>
+                // Start reading the next element in the queue we just took from.
+                completion.submit(completionTask(writerId))
+                r
+              case ReceiverEpochMarker(writerId) =>
+                // Don't read any more from this queue. If all the writers have sent epoch markers,
+                // the epoch is over; otherwise we need to poll from the remaining writers.
+                writerEpochMarkersReceived(writerId) = true
+                if (writerEpochMarkersReceived.forall(flag => flag)) {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: super nit: `writerEpochMarkersReceived.forall(_ == true)` is easier to understand.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala [hunk 14]
```diff
@@ -56,20 +69,71 @@ private[shuffle] class UnsafeRowReceiver(
...
+                  finished = true
+                  // Break out of the while loop and end the iterator.
+                  return null
+                } else {
+                  // Poll again for the next completion result.
+                  null
+                }
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: if you put `nextRow = newReceivedRow` inside the case ReceiverRow, then this else clause is not needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala (hunk 12 above)
Comment (tdas) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Actually ... finished is the flag, just use that for the while loop, that is, `while (!finished && nextRow == null)`.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala [hunk 15]
```diff
@@ -160,25 +170,122 @@ class ContinuousShuffleReadSuite extends StreamTest {
...
+          case _: InterruptedException => // do nothing - expected at test ending
+        }
       }
     }
 
     try {
       readRowThread.start()
       eventually(timeout(streamingTimeout)) {
-        assert(readRowThread.getState == Thread.State.WAITING)
+        assert(readRowThread.getState == Thread.State.TIMED_WAITING)
```
[the remark this answers is not in the document]
Comment (jose-torres) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: It changed because we added a timeout. The JVM always puts threads waiting without a timeout into WAITING, and threads waiting with a timeout into TIMED_WAITING.

[the remark this answers is not in the document]
Comment (tdas) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/shuffle/ContinuousShuffleReadSuite.scala]: I see. That's good to know!

**Comment by tdas** [comment]
LGTM! Will merge after tests pass.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala (hunk 7 above)
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/continuous/shuffle/UnsafeRowReceiver.scala]: Hi TD, just a question here, what's the 'a non-RPC-endpoint-based transfer mechanism' refers for?
